cpp/src/join/conditional_join_kernels.cuh

```diff
@@ -67,20 +67,21 @@ __global__ void compute_conditional_join_output_size(
     &intermediate_storage[threadIdx.x * device_expression_data.num_intermediates];
 
   std::size_t thread_counter{0};
-  cudf::size_type const start_idx      = threadIdx.x + blockIdx.x * block_size;
-  cudf::size_type const stride         = block_size * gridDim.x;
-  cudf::size_type const left_num_rows  = left_table.num_rows();
-  cudf::size_type const right_num_rows = right_table.num_rows();
-  auto const outer_num_rows            = (swap_tables ? right_num_rows : left_num_rows);
-  auto const inner_num_rows            = (swap_tables ? left_num_rows : right_num_rows);
+  auto const start_idx                         = cudf::detail::grid_1d::global_thread_id();
+  cudf::thread_index_type const stride         = thread_index_type{block_size} * gridDim.x;
+  cudf::thread_index_type const left_num_rows  = left_table.num_rows();
+  cudf::thread_index_type const right_num_rows = right_table.num_rows();
+  auto const outer_num_rows                    = (swap_tables ? right_num_rows : left_num_rows);
+  auto const inner_num_rows                    = (swap_tables ? left_num_rows : right_num_rows);
```

Comment:
The `auto` decls here look like an int64 type will be passed to the `evaluator.evaluate()` function which only accepts `size_type`. What happens in this case? Seems the compiler should've complained about this.

Reply:
I'll change the declarations of `[left|right]_row_index` below to perform the cast so that it's explicit. I don't think the compiler can warn in this case, narrowing larger integral types to smaller ones doesn't usually does it? At least from a few quick godbolt experiments I can't trigger an error without going from float->int or by introducing arithmetic operations that trigger overflow (at which point it's really warning about the overflow, not the narrowing).

Reply:
I thought I saw a warning for this kind of thing when compiled in debug mode but I also was not able to recreate it in godbolt. So I think your solution looks good.